kore/src/Kore/Step/Implication.hs

```diff
@@ -0,0 +1,452 @@
+mkImplication
+    :: modality
+    -> Pattern RewritingVariableName
+    -> OrPattern RewritingVariableName
+    -> [ElementVariable RewritingVariableName]
+    -> Implication modality
+mkImplication modality left right existentials =
+    Implication
+        { left
+        , right
+        , modality
+        , existentials
+        , attributes = Default.def
+        }
+
+{- | Construct a 'TermLike' from the parts of an implication-based rule.
+
+The 'TermLike' has the following form:
+
+@
+\\implies{S}(\\and{S}(left, requires), alias{S}(right))
+@
+
+that is,
+
+@
+left ∧ requires → alias(right)
+@
+
+ -}
+implicationToTerm
+    :: Modality
+    -> Implication modality
+    -> TermLike VariableName
+implicationToTerm modality representation@(Implication _ _ _ _ _) =
+    TermLike.mkImplies
+        (TermLike.mkAnd leftCondition leftTerm)
+        (TermLike.applyModality modality rightPattern)
+  where
+    Implication { left, right, existentials } = representation
+    leftTerm =
+        Pattern.term left
+        & getRewritingTerm
+    sort = TermLike.termLikeSort leftTerm
+    leftCondition =
+        Pattern.withoutTerm left
+        & Pattern.fromCondition sort
+        & Pattern.toTermLike
+        & getRewritingTerm
+    rightPattern =
+        TermLike.mkExistsN existentials (OrPattern.toTermLike right)
+        & getRewritingTerm
+
+substituteRight
+    :: Map
+        (SomeVariableName RewritingVariableName)
+        (TermLike RewritingVariableName)
+    -> Implication modality
+    -> Implication modality
+substituteRight rename implication'@Implication{ right, existentials } =
+    implication'
+        { right = OrPattern.substitute newSubst right
+        , existentials = renameVariable <$> existentials
+        }
+  where
+    existentials' = inject . TermLike.variableName <$> existentials
+    subst = foldr Map.delete rename existentials'
+    targetVars = Foldable.foldl' Set.union Set.empty
+        (FreeVariables.toNames . freeVariables @_ @RewritingVariableName
+        <$> subst)
+    freeVars = FreeVariables.toNames
+        (foldMap freeVariables (OrPattern.toPatterns right))
+        Set.\\ Map.keysSet subst
+        Set.\\ Set.fromList existentials'
+    avoid = Set.union targetVars freeVars
+    subst' = refreshVariables avoid (Set.fromList $ inject <$> existentials)
+    newSubst = Map.union subst (TermLike.mkVar <$> subst')
+
+    renameVariable
+        :: ElementVariable RewritingVariableName
+        -> ElementVariable RewritingVariableName
+    renameVariable var =
+        let name = SomeVariableNameElement . variableName $ var
+         in maybe var TermLike.expectElementVariable
+            $ Map.lookup name subst'
+
+renameExistentials
+    :: HasCallStack
+    => Map
+        (SomeVariableName RewritingVariableName)
+        (SomeVariable RewritingVariableName)
+    -> Implication modality
+    -> Implication modality
+renameExistentials rename implication'@Implication { right, existentials } =
```

Comment:
For efficiency reasons, mainly.
If we used this function, it would go ahead and do the variable renaming throughout `right` and we would then have to perform a new substitution across right in order to apply the `subst` that we're interested in.
In addition to that, I wasn't 100% convinced that the behaviour would be the same, but I can look into it if you feel like we should be using this function.